rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,301 @@
+
+```jsx
+<ClientMessage type="missingValue">Please provide a name.</ClientMessage> // will yield "Please provide a name."
+```
+
+#### Client-side validation types
+
+`Form.ClientMessage` accepts a required `type` prop which is used to determine when the message should show. It matches the native HTML validity state (`ValidityState` on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState)) which result from using attributes such as `required`, `min`, `max`. The message will show if the given `type` is `true` on the control’s validity state.
+
+```ts
+// `type` is one of the following keys from the DOM ValidityState interface:
+
+interface ValidityState {
+  readonly badInput: boolean;
+  readonly customError: boolean;
+  readonly patternMismatch: boolean;
+  readonly rangeOverflow: boolean;
+  readonly rangeUnderflow: boolean;
+  readonly stepMismatch: boolean;
+  readonly tooLong: boolean;
+  readonly tooShort: boolean;
+  readonly typeMismatch: boolean;
+  readonly valid: boolean;
+  readonly valueMissing: boolean;
+}
+```
+
+This means you can even show something whe the field is valid:
+
+```jsx
+<Form.ClientMessage type="valid">✅ Valid!</Form.ClientMessage>
+```
+
+#### Custom validation
+
+On top of all the built-in client-side validation types described above you can also provide your own custom validation whislt still making use of the platform's validation abilities. It uses the `customError` type present in the constraint validity API.
+
+You can pass `type="customError"` on `Form.ClientMessage` and provide an `isValid` prop with your custom validation logic.
+
+> `isValid` is called with the current value of the control and the current values of all other controls in the form.
+
+Here's a contrived example:
+
+```jsx
+<Form.Field name="name">
+  <Form.Label>Full name</Form.Label>
+  <Form.Control />
+  <Form.ClientMessage type="customError" isValid={(value, fields) => value === 'John'}>
+    Only John is allowed.
+  </Form.ClientMessage>
+</Form.Field>
+```
+
+> `isValid` can also be an `async` function (or return a promise) to perform async validation (see type below):
+
+```ts
+type CustomValidatorFn = ValidatorSyncFn | ValidatorAsyncFn;
+type ValidatorSyncFn = (value: string, fields: FormFields) => boolean;
+type ValidatorAsyncFn = (value: string, fields: FormFields) => Promise<boolean>;
+type FormFields = { [index in string]?: FormDataEntryValue };
+```
+
+#### Accessing the validity state for even more control
+
+Sometimes, you may need to access the raw validity state of a field in order to display your own icons, or interface with your own component library by passing certain props to it. You can do this by using the `Form.ValidityState` part:
+
+```jsx
+<Form.Field name="name">
+  <Form.Label>Full name</Form.Label>
+  <Form.ValidityState>
+    {(validity) => (
+      <Form.Control asChild>
+        <TextField variant="primary" state={getTextFieldState(validity)} />
+      </Form.Control>
+    )}
```

Comment:
Radix got so far without having to resort to render props — why add this pattern now?

Couldn't it be a component instead?

```tsx
<Form.ValidityState value="success">
  ... success state ...
</Form.ValidityState>
```

Reply:
Hey Paco, sure a component-based API could be created to render based on certain state matching (pretty much a simpler version of `ClientMessage`) but sometimes you do need more access.

ie. the case you highlighed above where your DS expects a prop for some variant or state.

Worth nothing also that although we we haven't historically used render props, we are not completely opposed to it. It depends what the use-case is and if there are other technical solution with less API surface. For most of the states, we do believe a controlled prop fulfills the need.

With validity though, that isn't something that can be "controlled". However, we could follow the same logic and expose a callback for when the validity of a field changes so you have access to it potentially?

I do feel like a render prop is a good fit for this particular case though.

Do you have specific concerns about it?